python/server/qgsserverresponse.sip

```diff
@@ -0,0 +1,80 @@
+/***************************************************************************
+                          qgsserverresponse.sip
+
+  Define response class for services
+  -------------------
+  begin                : 2016-12-05
+  copyright            : (C) 2016 by David Marteau
+  email                : david dot marteau at 3liz dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+/**
+ * \ingroup server
+ * QgsServerResponse
+ * Class defining response interface passed to services QgsService::executeRequest() method
+ */
+
+class QgsServerResponse
+{
+%TypeHeaderCode
+#include "qgsserverresponse.h"
+%End
+    public:
+       
+        //!constructor
+        QgsServerResponse();
+
+        //! destructor
+        virtual ~QgsServerResponse();
+
+       /** Set Header entry 
+        *  Add Header entry to the response 
+        *  Note that it is usually an error to set Header after writng data
+        */
+        virtual void setHeader( const QString& key, const QString& value ) = 0;
+        
+       /** Set the http return code
+        * @param code HTTP return code value
+        */
+        virtual void setReturnCode( int code ) = 0;
+
+        /**
+         * Send error
+         * This method delegate error handling at the server level. This is different
+         * from calling setReturnCodei() along with and a specific response body.
+         * @param code HHTP return code value
+         * @param message An informative error message
+         */
+        virtual void sendError( int code,  const QString& message ) = 0;
```

Comment:
@mhugo you can return legitimate response body with different return code (ex: 200 to 206 codes) so you  should have the possibility to set a return code wich does not necessary indicates an error. Also, you should be able to return a true error code without having to deal with the particular format of the error response which must be handled uniformly across all services.